File: frontend/src/api/dto/rideTemplateDto.ts
import type { Instant } from './instant.ts'
import type { RideTemplateGroupDto } from './rideTemplateGroupDto.ts'
import type { Status } from './status.ts'
import type { TeamPublicationDto } from './teamPublicationDto.ts'
import type { Visibility } from './visibility.ts'

/**
 * Ride template response
 */
export interface RideTemplateDto {
  /** Team */
  team: TeamPublicationDto
  /** Template ID (TSID) */
  id: string
  /** Template slug */
  slug: string
  /** Template name */
  name: string
  /** Template description (markdown) */
  markdown: string
  /** Visibility level */
  visibility: Visibility
  /** Default status */
  status: Status
  /** Creation timestamp */
  createdAt: Instant
  /** Last update timestamp */
  updatedAt: Instant
  /** Number of groups */
  groupCount: number
  /** Template groups */
  groups: RideTemplateGroupDto[]
}

File: backend/src/main/java/fr/pedalons/dto/ridetemplates/response/RideTemplateDto.java
package fr.pedalons.dto.ridetemplates.response;

import fr.pedalons.common.TsidUtils;
import fr.pedalons.domain.ridetemplate.RideTemplate;
import fr.pedalons.dto.publications.response.TeamPublicationDto;
import fr.pedalons.dto.validation.ValidateSchema;
import fr.pedalons.enums.Status;
import fr.pedalons.enums.Visibility;
import java.time.Instant;
import java.util.List;
import org.eclipse.microprofile.openapi.annotations.media.Schema;

@Schema(description = "Ride template response")
@ValidateSchema
public record RideTemplateDto(
    @Schema(description = "Team", required = true) TeamPublicationDto team,
    @Schema(description = "Template ID (TSID)", required = true) String id,
    @Schema(description = "Template slug", required = true) String slug,
    @Schema(description = "Template name", required = true) String name,
    @Schema(description = "Template description (markdown)", required = true) String markdown,
    @Schema(description = "Visibility level", required = true) Visibility visibility,
    @Schema(description = "Default status", required = true) Status status,
    @Schema(description = "Creation timestamp", required = true) Instant createdAt,
    @Schema(description = "Last update timestamp", required = true) Instant updatedAt,
    @Schema(description = "Number of groups", required = true) int groupCount,
    @Schema(description = "Template groups", required = true) List<RideTemplateGroupDto> groups) {

  public static RideTemplateDto from(RideTemplate template) {
    List<RideTemplateGroupDto> groupDtos =
        template.getGroups().stream().map(RideTemplateGroupDto::from).toList();

    return new RideTemplateDto(
        TeamPublicationDto.from(template.getTeam()),
        TsidUtils.toString(template.getId()),
        template.getSlug(),
        template.getName(),
        template.getMarkdown(),
        template.getVisibility(),
        template.getStatus(),
        template.getCreatedAt(),
        template.getUpdatedAt(),
        groupDtos.size(),
        groupDtos);
  }
}
